uber/site_sections/departments.py

```diff
@@ -0,0 +1,218 @@
+
+        if cherrypy.request.method == 'POST':
+            department = session.department(
+                params,
+                bools=Department.all_bools,
+                checkgroups=Department.all_checkgroups)
+            message = check(department)
+            if not message:
+                session.add(department)
+                raise HTTPRedirect('form?id={}', department.id)
+        else:
+            department = session.query(Department).options(
+                subqueryload(Department.dept_roles)
+                    .subqueryload(DeptRole.dept_memberships),
+                subqueryload(Department.members)
+                    .subqueryload(Attendee.shifts)
+                        .subqueryload(Shift.job),
+                subqueryload(Department.unassigend_requesting_attendees),
+                subqueryload(Department.dept_heads)
+                    .subqueryload(Attendee.dept_memberships),
+                subqueryload(Department.pocs)
+                    .subqueryload(Attendee.dept_memberships),
+                subqueryload(Department.checklist_admins)
+                    .subqueryload(Attendee.dept_memberships)
+                ).get(department_id)
+
+        return {
+            'admin': session.admin_attendee(),
+            'message': message,
+            'department': department
+        }
+
+    @requires_dept_admin
+    @csrf_protected
+    def delete(self, session, id, message=''):
+        if cherrypy.request.method == 'POST':
+            department = session.query(Department).get(id)
+            attendee = session.admin_attendee()
+            if department.member_count > 1:
+                raise HTTPRedirect(
+                    'form?id={}&message={}',
+                    id,
+                    'You cannot delete a department with more than one member')
+
+            session.delete(department)
+            raise HTTPRedirect(
+                'index?message={}',
+                'The {} department was deleted'.format(department.name))
+
+        raise HTTPRedirect('form?id={}', id)
+
+    def new(self, session, message='', **params):
+        if params.get('id'):
+            raise HTTPRedirect('form?id={}', params['id'])
+
+        department = session.department(
+            params,
+            bools=Department.all_bools,
+            checkgroups=Department.all_checkgroups)
+
+        if cherrypy.request.method == 'POST':
+            message = check(department)
+            if not message:
+                attendee = session.admin_attendee()
+                has_email = bool(attendee.email)
+                department.memberships = [DeptMembership(
+                    attendee=attendee,
+                    is_dept_head=True,
+                    is_poc=has_email,
+                    is_checklist_admin=has_email)]
+                session.add(department)
+                raise HTTPRedirect('form?id={}', department.id)
+            session.rollback()
```

Comment:
Why the rollback here?  We don't generally need to do that in our forms, and I don't see a change that specifically needs to be rolled back offhand.

Reply:
Ah, yes, that rollback is not necessary. At one point I was setting foreign relations on the new object, which would incidentally add it to the session. No longer needed, though!